src/platforms/hosted/jlink.c

```diff
@@ -277,6 +234,313 @@ static bool jlink_claim_interface(void)
 	return true;
 }
 
+/* J-Link command functions and utils */
+
+static char *jlink_hw_type_to_string(const uint8_t hw_type)
+{
+	switch (hw_type) {
+	case JLINK_HARDWARE_VERSION_TYPE_JLINK:
+		return "J-Link";
+	case JLINK_HARDWARE_VERSION_TYPE_JTRACE:
+		return "J-Trace";
+	case JLINK_HARDWARE_VERSION_TYPE_FLASHER:
+		return "Flasher";
+	case JLINK_HARDWARE_VERSION_TYPE_JLINKPRO:
+		return "J-Link Pro";
+	default:
+		return "Unknown";
+	}
+}
+
+static char *jlink_interface_to_string(const uint8_t interface)
+{
+	switch (interface) {
+	case JLINK_INTERFACE_JTAG:
+		return "JTAG";
+	case JLINK_INTERFACE_SWD:
+		return "SWD";
+	case JLINK_INTERFACE_BDM3:
+		return "BDM3";
+	case JLINK_INTERFACE_FINE:
+		return "FINE";
+	case JLINK_INTERFACE_2W_JTAG_PIC32:
+		return "PIC32 2-wire JTAG";
+	case JLINK_INTERFACE_SPI:
+		return "SPI";
+	case JLINK_INTERFACE_C2:
+		return "C2";
+	case JLINK_INTERFACE_CJTAG:
+		return "cJTAG";
+	default:
+		return "Unknown";
+	}
+}
+
+static bool jlink_get_version(void)
+{
+	uint8_t buffer[4U];
+
+	/* Read the firmware version, replies with a 2 byte length packet followed by the version string packet */
+	if (!jlink_simple_query(JLINK_CMD_INFO_GET_FIRMWARE_VERSION, buffer, 2U))
+		return false;
+
+	/* Verify the version string fits in the buffer (expected value is 0x70) */
+	const uint16_t version_length = read_le2(buffer, 0);
+	if (version_length > sizeof(jlink.fw_version))
+		return false;
+
+	/* Read vesion string directly into jlink.version */
+	bmda_usb_transfer(info.usb_link, NULL, 0, jlink.fw_version, version_length, JLINK_USB_TIMEOUT);
+	jlink.fw_version[version_length - 1U] = '\0'; /* Ensure null termination */
+
+	DEBUG_INFO("Firmware version: %s\n", jlink.fw_version);
+
+	/* Read the hardware version if supported */
+	if (jlink.capabilities & JLINK_CAPABILITY_HARDWARE_VERSION) {
+		if (!jlink_simple_query(JLINK_CMD_INFO_GET_HARDWARE_VERSION, buffer, 4U))
+			return false;
+
+		jlink.hw_version = read_le4(buffer, 0);
+
+		DEBUG_INFO("Hardware Version: %s V%u.%u.%u\n",
+			jlink_hw_type_to_string(JLINK_HARDWARE_VERSION_TYPE(jlink.hw_version)),
+			JLINK_HARDWARE_VERSION_MAJOR(jlink.hw_version), JLINK_HARDWARE_VERSION_MINOR(jlink.hw_version),
+			JLINK_HARDWARE_VERSION_REVISION(jlink.hw_version));
+	}
+
+	return true;
+}
+
+static bool jlink_get_capabilities(void)
+{
+	uint8_t buffer[4U];
+	if (!jlink_simple_query(JLINK_CMD_INFO_GET_PROBE_CAPABILITIES, buffer, sizeof(buffer)))
+		return false;
+
+	jlink.capabilities = read_le4(buffer, 0);
+	DEBUG_INFO("Capabilities: 0x%" PRIx32 "\n", jlink.capabilities);
+
+	return true;
+}
+
+static bool jlink_interface_available(const uint8_t interface)
```

Comment:
We'd suggest `static inline bool` given this is a one-liner and trivially inlined